TUTORIAL.md

```diff
@@ -0,0 +1,245 @@
+  
+    <label for="post_tags_attributes_2_name">Name</label>
+    <input type="text" name="post[tags_attributes][2][name]" id="post_tags_attributes_2_name" />
+  </div>
+
+  <div class="actions">
+    <input type="submit" name="commit" value="Create Post" />
+  </div>
+
+</form>
+```
+
+In the first section of our form, we can see our form helper generated `post[name]` and `post[content]`. This is great, we have access to our post name and content.
+
+#### `fields_for`
+Let's focus on our tag fields for a second, how did Rails know to create such a nicely nested structure?
+
+If we look at our form, we will see `<%= f.fields_for(:tags) do |tag_form| %>`. Our `fields_for` accepts a `:tags` object and by convention introspects into our `Post` model to see if it has a way to accept this new object. It is specifically looking for a macro called `accepts_nested_attributes_for :tags`. If it finds it, it will generate the nicely nested form that we see above.
+
+In order to understand `accepts_nested_attributes_for`, we need to take a look at our models associations and the methods that were made available to us.
+
+***note that we have added `accepts_nested_attributes_for :tags` to our `Post` model**
+
+##Step 4: Set up Post model
+### Models
+
+```ruby
+class Post < ActiveRecord::Base
+  has_many :post_tags
+  has_many :tags, :through => :post_tags
+
+  accepts_nested_attributes_for :tags
+end
+
+class PostTag < ActiveRecord::Base
+  belongs_to :post
+  belongs_to :tag
+end
+
+class Tag < ActiveRecord::Base
+  has_many :post_tags
+  has_many :posts, :through => :post_tags
+end
+```
+By creating these associations, Rails gives us the methods below.
+
+```ruby
+tag_ids()                    Post (Post::GeneratedAssociationMethods)
+tag_ids=(ids)                Post (Post::GeneratedAssociationMethods)
+tags(*args)                  Post (Post::GeneratedAssociationMethods)
+tags=(value)                 Post (Post::GeneratedAssociationMethods)
+tags_attributes=(attributes) Post (Post::GeneratedAssociationMethods)
+```
+What we are interested in right now is `tags_attributes=(attributes)`
+
+As mentioned before, our `fields_for` is looking for `accepts_nested_attributes_for :tags`. But what exactly does this macro give us? Let's define our own and see.
+
+```ruby
+def tags_attributes=(attributes_hash)
+  attributes_hash.values.each do |attributes|
+    self.tags.build(attributes)
+  end
+end
+```
+If we create three new tags and look at params in our console, we will see the following in our `attributes_hash`.
+
+```ruby
+{
+    "0" => {
+        "name" => "tag1"
+    },
+    "1" => {
+        "name" => "tag2"
+    },
+    "2" => {
+        "name" => "tag3"
+    }
+}
+```
+Great! It looks like our `tags_attributes=` method is iterating over our new tags, grabbing the values and building our tags. Then when `@post` is saved, it calls our `tag_attributes=` method in the `Post` model.
```

Comment:
I would frame this a bit differently. We already have the `build` method. That method, can take some arguments. Once you understand what arguments the build method can take in to create a tags object. Then you need to figure out how to get those arguments. So let's see what we get back in our `params` that we can then pass into `build`. Oh look! we have a bunch of stuff athat looks good. How do we get into all of those things? Oh it's the values. done.